### PR TITLE
[CORE] Now traces process creation

### LINKED ISSUE #101: No process relationship is recorded
Running the first test in the tests directory, it reports processes:

> p0 a b:process .
> p0 b:cmd "gcc -c f1.c" .
> p1      a       b:process .
> p1      b:cmd   "/usr/lib/gcc/x86_64-linux-gnu/11/cc1 -quiet -imultiarch x86_64-linux-gnu f1.c -quiet -dumpbase f1.c -dumpbase-ext .c -mtune=generic -march=x86-64 -fasynchronous-unwind-tables -fstack-protector-strong -Wformat -Wformat-security -fstack-clash-protection -fcf-protection -o /tmp/ccDu6HTo.s" .
> p2      a       b:process .
> p2      b:cmd   "as --64 -o f1.o /tmp/ccDu6HTo.s" .

but no relationship between them (i.e. p0 forked p1)

### PR DESCRIPTION
**Pull request by fvalasiad**
Closes #101

### CODE REVIEW

File: src/tracer.c [hunk 1]
```diff
@@ -227,8 +227,18 @@ handle_syscall(PROCESS_INFO *pi, const struct ptrace_syscall_info *entry,
 	    // int flags);
 	    record_process_start(pi->pid, pi->outname);
 	    break;
-	default:
```
Comment (zvr) [inline src/tracer.c:230]: Why is this removed?

File: src/tracer.c [hunk 2]
```diff
@@ -227,8 +227,18 @@ handle_syscall(PROCESS_INFO *pi, const struct ptrace_syscall_info *entry,
...
-	    return;
+	case SYS_fork:
+	    // pid_t fork(void);
+	    record_process_create(pi->outname, find(exit->exit.rval)->outname);
```
Comment (zvr) [inline src/tracer.c:232]: How do you expect to `find()` the new process? Where was it entered in `pinfo` ?

**Review by zvr — CHANGES_REQUESTED** [review]

File: src/tracer.c (hunk 2 above)
Comment (fvalasiad) [reply src/tracer.c:232]: @zvr The order of stops upon one of those syscalls is:

first stop: `PTRACE_SYSCALL_ENTRY`
second stop: stop from setting `PTRACE_O_TRACEFORK`, `PTRACE_O_TRACEVFORK`, `PTRACE_O_TRACECLONE`
last stop: `PTRACE_SYSCALL_EXIT`


Is that documented anywhere? Perhaps it is.
`ptrace(2)`'s manual states:

**Stop the tracee at the next fork(2) and automatically
                     start tracing the newly forked process, which will start
                     with a SIGSTOP, or PTRACE_EVENT_STOP if PTRACE_SEIZE was
                     used.  A waitpid(2) by the tracer will return a status
                     value such that

                       status>>8 == (SIGTRAP | (PTRACE_EVENT_FORK<<8))

                     The PID of the new process can be retrieved with
                     PTRACE_GETEVENTMSG.**
                
Now you might argue that this doesn't assume that the stop is caused before `fork(2)` exits. But i am pretty sure the existence of this alternative option does:

**PTRACE_O_TRACEVFORKDONE (since Linux 2.5.60)
                     Stop the tracee at the completion of the next vfork(2).
                     A waitpid(2) by the tracer will return a status value
                     such that

                       status>>8 == (SIGTRAP | (PTRACE_EVENT_VFORK_DONE<<8))

                     The PID of the new process can (since Linux 2.6.18) be
                     retrieved with PTRACE_GETEVENTMSG.**
                     
And thus i made the assumption that the new process will not only exist, but it will also be at the top of the array(the last element). Althought taking into considerations that there might be some weird scenarios where this isn't the case, i just used `find(3)`.

File: src/tracer.c (hunk 1 above)
Comment (fvalasiad) [reply src/tracer.c:230]: @zvr Because it's redundant, after the switch the function exits anyway. Do we care about keeping it?

File: src/tracer.c (hunk 2 above)
Comment (fvalasiad) [reply src/tracer.c:232]: @zvr `tracer_main(3)` at the `SIGSTOP` switch case.

Comment (fvalasiad) [reply src/tracer.c:232]: And oh well why did i add those three syscall handlers instead of handling it at `SIGSTOP` you might ask? Well it's because we can't really know the one who created the new process at the `SIGSTOP` switch case.

What we could do tho is use the fact that the one who created the new process is also stopped with a `SIGTRAP` as stated in the manual and add a handler in `tracer_main` that does just that. Would we prefer that?
